### PR TITLE
quay: introduce a quay.io client library

### PR DESCRIPTION
**Pull request by lucab**
This introduces an asynchronous quay.io client, based on reqwest.
Its feature set is currently limited to fetching tags and labels.

### CODE REVIEW

**Comment by lucab** [comment]
/cc @steveeJ 
This should already implement the minimal logic required by cincinnati to scrape labels on active tags. Authentication logic is as described [here](https://docs.quay.io/api/#generating-a-token-for-internal-application-use) but I haven't tested it yet. CI will need an access token for testing too.

**Comment by cgwalters** [comment]
I'll just say I skimmed this, looks sane to me.  I found it useful to read Futures code as I don't have much experience with it.

**Comment by lucab** [comment]
@steveeJ I fixed and validated the authentication part too, manually on a personal private organization. Test for private images are written but not enabled (pending a CI token), some of the assertions may be slightly off. I left a TODO for pagination. PTAL.

**Comment by steveej** [comment]
This is quite some code, thanks for the effort! Due to the size of this I will start integrating it with the rest of the code and see how it performs there. I will add my review once that's done.

**Comment by steveej** [comment]
Oh, I missed that functionality as I was looking for a "label" file. So the
*labels.rs* can be dropped from my commit. Do you mind if I force push to
this branch?

**Comment by lucab** [comment]
@steveeJ feel free to force push to this, I locally duplicated my stuff in another private branch.

**Comment by steveej** [comment]
@lucab 

> @steveeJ feel free to force push to this, I locally duplicated my stuff in another private branch.

I merged the additional code with the `manifest.rs`. I had also done a rebase but feel free to just cherry-pick my two commits on top of a local branch and re-force-push ;-)

File: quay/src/v1/mod.rs
Comment (steveej) [inline quay/src/v1/mod.rs:59]: I would add a `set_` or `with_` prefix to this method name because it usually means to get the thing if it doesn't say otherwise.

File: quay/src/v1/mod.rs
Comment (steveej) [inline quay/src/v1/mod.rs:66]: I would add a `set_` or `with_` prefix to this method name because it usually means to get the thing if it doesn't say otherwise.

File: quay/src/v1/mod.rs
Comment (steveej) [inline quay/src/v1/mod.rs:73]: I would add a `set_` or `with_` prefix to this method name because it usually means to get the thing if it doesn't say otherwise.

**Review by steveej — APPROVED** [review]
Optionally you could add a `use tokio::runtime::Runtime` in the tests as it's used quite often.

**Comment by steveej** [comment]
PTAL at the comments before we merge this. I consider them optional so just remove the hold or ping me back after you addressed them.

/lgtm
/hold

File: quay/src/v1/mod.rs
Comment (lucab) [reply quay/src/v1/mod.rs:59]: The advice is generally true. However in this specific case this a builder, and in my experience so far builders don't explicitly label methods as setters (I guess because they don't have getters):
 * https://doc.rust-lang.org/std/fs/struct.DirBuilder.html
 * https://doc.rust-lang.org/std/fs/struct.OpenOptions.html
 * https://docs.rs/reqwest/0.9.9/reqwest/struct.ClientBuilder.html

I'll leave these methods without the prefix (same for the following comments too).

**Comment by lucab** [comment]
@steveeJ I've added use statements (but stuck with current names) and repushed. From my side this is ok to land.

/lgtm

**Pull request merged**
